c10/cuda/CUDACachingAllocator.cpp

```diff
@@ -772,8 +723,7 @@ class THCCachingAllocator {
     for (auto it = streams.begin(); it != streams.end(); ++it) {
       C10_CUDA_CHECK(cudaSetDevice(it->device_index()));
 
-      cudaEvent_t event;
-      C10_CUDA_CHECK(cudaEventCreateWithFlags(&event, cudaEventDisableTiming));
+      cudaEvent_t event = create_event_internal();
```

Comment:
> is there anything about the cudaEvent_t internals that would favor (or disfavor) caching them in the PyTorch allocator? For reference they're currently created and recorded here (due to a prior call to record_stream):

I think there's a subtle problem with event-caching here.  If a block is allocated on one device and used on a stream on another device (ie if a kernel on dev 1 pulls from the memory on dev 0) then the block on dev 0 could get a stream on dev 1 recorded to it.  Line 740 just above (`(C10_CUDA_CHECK(cudaSetDevice(it->device_index()));`) anticipates the possibility that `it->stream()` may be on a different device.

["cudaEventRecord() will fail if the input event and input stream are associated to different devices."](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#stream-and-event-behavior).  The device associated with an event is the device that's current at the time [cudaEventCreateWithFlags](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__EVENT.html#group__CUDART__EVENT_1g7b317e07ff385d85aa656204b971a042) is called.  In the pre-PR code path, there's no problem because the event is created on the fly after `cudaSetDevice`, so it's guaranteed to be associated with `it->stream()`'s device.  However, if you're pulling an event from the cache, there's no telling what device was current when it was `cudaEventCreate`ed.

I asked Cuda devs internally
> Is it safe to regard cudaEventCreate/Destroy as relatively cheap?  Is creating events on the fly, recording them and using them for a single wait, then destroying them a reasonable pattern?

and they said
> With cudaEventDisableTiming it should be "cheap", as there is no extra state needed to be allocated....However, there are still malloc [regular CPU ones] or two on the path.

What's your motivation for caching events?  The stream recording/event processing logic is only triggered for cross-stream memory use, which is an uncommon case, so the cache may introduce brittleness to serve an "optimization" we don't really need.  Does event caching improve performance substantially for cross-stream memory use?

Good news is, it appears event caching is orthogonal to what this PR wants to do, and it will be easy to drop in the old on-the-fly event pattern here instead.

Reply:
Good catch.  I'll remove that part.